### ISSUE TITLE
Does the shaper _need_ to return an error?

### ISSUE THREAD

**Issue by whereswaldon** [go-text/typesetting#27]
So the `shaping.Shaper` interface has this definition (as of #25):

```go
type Shaper interface {
	// Shape takes an Input and shapes it into the Output.
	Shape(Input) (Output, error)
}
```

There are only two ways for that error case to happen though:

1. [invalid run specification](https://github.com/go-text/typesetting/blob/main/shaping/shaper.go#L56)
1. [harfbuzz returning a GID that doesn't exist](https://github.com/go-text/typesetting/blob/main/shaping/shaper.go#L88)

It seems like 1. could be solved by choosing a safe interpretation (like swapping so start < end) and 2. may be entirely hypothetical. If that's the case, I think we should drop the error return.

@benoitkugler, is 2. entirely hypothetical? I wrote that code because the GlyphExtents method has a validity return value, but it seems kinda unlikely or impossible for that to come into play here. Additionally, it would be relatively safe to just set the relevant values to zero if that call failed.

Having this signature no longer return an error dramatically increases ergonomics, so I think we should strive for that if it's possible.

**Comment by benoitkugler** [comment]
> @benoitkugler, is 2. entirely hypothetical? I wrote that code because the GlyphExtents method has a validity return value, but it seems kinda unlikely or impossible for that to come into play here. Additionally, it would be relatively safe to just set the relevant values to zero if that call failed

Good question, with no simple answer. Harfbuzz tries its best to use glyph indeed present in the font, but it cannot actually ensure it because of glyph substitutions. I've a few exemple in the test suite that produce invalid glyphs.
However, I think that in the large majority of the use cases, glyphs will be valid. Thus, handling the few edge cases by removing the glyph (or zeroing its width) seems a good idea.
The thing is that the end user won't be able to act on the error anyway since it will probably come from the font file, on which he often has not much control...

Overall, I'm also in favor of dropping the error, and doing a best effort to handle the very rare edge cases.

**Comment by whereswaldon** [comment]
Okay, agreed. We can try to tackle this change once #25 merges.

**Comment by whereswaldon** [comment]
Resolved by #25
